src/Sulu/Component/HttpCache/EventSubscriber/KernelSubscriber.php

```diff
@@ -0,0 +1,111 @@
+<?php
+/*
+ * This file is part of the Sulu CMS.
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\HttpCache\EventSubscriber;
+
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Sulu\Component\HttpCache\StructureCacheHandlerInterface;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\Event\PostResponseEvent;
+use Sulu\Component\HttpCache\HandlerInterface;
+
+/**
+ * Listen to the content mapper and invalidate structures
+ */
+class KernelSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @var Sulu\Component\HttpCache\StructureCacheHandlerInterface
+     */
+    private $handler;
+
+    /**
+     * @var Sulu\Component\Content\StructureInterface
+     */
+    private $structure;
+
+    /**
+     * @var boolean
+     */
+    private $doFlush = false;
+
+    /**
+     * {@inheritDoc}
+     */
+    public static function getSubscribedEvents()
+    {
+        return array(
+            KernelEvents::RESPONSE => 'onResponse',
+            KernelEvents::REQUEST => 'onRequest',
+            KernelEvents::TERMINATE => 'onTerminate',
+        );
+    }
+
+    /**
+     * @param StructureCacheHandlerInterface $handler
+     */
+    public function __construct(HandlerInterface $handler)
+    {
+        $this->handler = $handler;
+    }
+
+    /**
+     * Get the Structure from the main request for use later.
+     *
+     * @param ContentNodeEvent $event
+     */
+    public function onRequest(GetResponseEvent $event)
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $request = $event->getRequest();
+
+        if (false === $request->attributes->has('structure')) {
+            return;
+        }
+
+        // do not cache preview
+        if (true === $request->query->has('preview')) {
+            return;
+        }
+
+        $this->structure = $request->attributes->get('structure');
+    }
+
+    /**
+     * If the structure has been picked up from the request, updated
+     * the response using the configured handler
+     *
+     * @param FilterResponseEvent
+     */
+    public function onResponse(FilterResponseEvent $event)
+    {
+        if (null === $this->structure) {
+            return;
+        }
+
+        $this->handler->updateResponse($event->getResponse(), $this->structure);
+        $this->doFlush = true;
+    }
+
+    /**
+     * Flush the cache on kernel terminate
+     *
+     * @param PostResponseEvent $event
+     */
+    public function onTerminate(PostResponseEvent $event)
+    {
+        $this->handler->flush();
```

Comment:
It still makes sense to have the HandlerInterface. Handlers may implement one or more of the Handler interfaces, but they are all handlers.

The validity of this design pattern used here is debatable, but its too late to change imo.